esp/services/esdl_svc_engine/esdl_svc_custom.cpp

```diff
@@ -0,0 +1,331 @@
+                compileNestedClauses(otherwise, true);
+            }
+        }
+        else
+            ERRLOG("CEsdlTransformConditional: Found xsl:choose clause without required xsl:when");
+    }
+}
+
+void CEsdlCustomTransformClause::compileNestedClauses(IPropertyTree * nested, bool otherwise)
+{
+    Owned<IPropertyTreeIterator> conditionalIterator = nested->getElements("xsl:choose");
+    ForEach(*conditionalIterator)
+    {
+        auto xslchooseelement = &conditionalIterator->query();
+        addNestedTransformClause(new CEsdlCustomTransformClause(xslchooseelement), otherwise);
+    }
+}
+
+void CEsdlCustomTransformClause::compileRules(IPropertyTreeIterator * rulesiter, bool otherwise, bool typeIsSet)
+{
+    ForEach(*rulesiter)
+    {
+        IPropertyTree & cur = rulesiter->query();
+
+        bool required = cur.getPropBool("@required", false);
+        const char * ruleName = cur.queryProp("@name");
+        const char * targetFieldPath = cur.queryProp("hpcc:TargetField/@path");
+        if (!targetFieldPath || !*targetFieldPath)
+        {
+            //if(required)
+                //throw??
+            //else
+            ERRLOG("Encountered SetValue without TargetField path");
+            continue;
+        }
+
+        const char * pathToValue = cur.queryProp("hpcc:Value/@path");
+        if (!pathToValue || !*pathToValue)
+        {
+            //if(required)
+               //throw??
+            //else
+            ERRLOG("Encountered SetValue without Value path");
+            continue;
+        }
+
+        addTransformRule(new CEsdlCustomTransformRule(ruleName, targetFieldPath, pathToValue, required, typeIsSet ), otherwise);
+    }
+}
+
+void CEsdlCustomTransformClause::compileRules(IPropertyTree * rules, bool otherwise)
+{
+    compileRules(rules->getElements("hpcc:SetValue"), otherwise, true);
+    compileRules(rules->getElements("hpcc:AppendValue"), otherwise, false);
+}
+
+void CEsdlCustomTransformClause::addTransformRule(CEsdlCustomTransformRule * fieldmapping, bool otherwise)
+{
+    if (!otherwise)
+        m_transformRules.append(*LINK(fieldmapping));
+    else
+        m_otherwiseTransformRules.append(*LINK(fieldmapping));
+}
+
+void CEsdlCustomTransformClause::addNestedTransformClause(CEsdlCustomTransformClause * nestedConditional, bool otherwise)
+{
+    if(!otherwise)
+        m_nestedTransformClauses.append(*LINK(nestedConditional));
+    else
+        m_otherwiseNestedTransformClauses.append(*LINK(nestedConditional));
+}
+
+bool CEsdlCustomTransformClause::evaluate(IEspContext * context, IPropertyTree *request, IPropertyTree * bindingCfg)
+{
+    StringBuffer xmlreq;
+    toXML(request, xmlreq);
+
+    m_xpathContext->setXmlDoc(xmlreq.str());
+
+    VStringBuffer ver("%g", context->getClientVersion());
+    if(!m_xpathContext->addParameter("clientversion", ver.str()))
+        ERRLOG("Could not set custom transform variable: clientversion:'%s'", ver.str());
+
+    auto user = context->queryUser();
```

Comment:
@timothyklemm  please comment on the use of the getPropertyIterator you exposed.

Reply:
getPropertyIterator returns a new iterator instance, or NULL when no properties have been set. The returned value should either be 'Owned' or explicitly released. It must not be dereferenced without first checking for NULL.